src/graph_compute/update_function.h

```diff
@@ -80,30 +121,99 @@ typedef struct vertex_t vertex_t;
     WHEN_TEST({
       __sync_add_and_fetch(&roundUpdateCount, 1);
     })
-    vertex_t * current = &nodes[index];
-    // recalculate this node's data
+  #if IN_PLACE
+    data_t * next = &nodes[index].data;
+  #else
+    data_t * next = &nodes[index].data[(round + 1) & 1];
+  #endif
+
+    // recalculate this node's pagerank
     pagerank_t pagerank = 0;
-    for (vid_t i = 0; i < current->cntEdges; ++i) {
+    for (vid_t i = 0; i < nodes[index].cntEdges; i++) {
     #if IN_PLACE
-      pagerank += nodes[current->edges[i]].data.contrib;
+      pagerank += nodes[nodes[index].edges[i]].data.contrib;
     #else
-      pagerank += nodes[current->edges[i]].data[round & 1].contrib;
+      pagerank += nodes[nodes[index].edges[i]].data[round & 1].contrib;
     #endif
     }
+    pagerank *= globaldata->d;
+    pagerank += (1 - globaldata->d);
+  #if TEST_CONVERGENCE
   #if IN_PLACE
-    #if TEST_CONVERGENCE
-      nodes[index].data.pagerankOld = nodes[index].data.pagerank;
-    #endif
-    nodes[index].data.pagerank = pagerank;
-    nodes[index].data.contrib = pagerank/static_cast<pagerank_t>(current->cntEdges);
+    data_t * current = &nodes[index].data;
   #else
-    nodes[index].data[(round+1) & 1].pagerank = pagerank;
-    nodes[index].data[(round+1) & 1].contrib =
-      pagerank/static_cast<pagerank_t>(current->cntEdges);
+    data_t * current = &nodes[index].data[round & 1];
+  #endif
+    pagerank_t averageDiff = std::abs(pagerank - current->pagerank);
+    globaldata->averageDiff[round] += averageDiff;
+    assert(PARALLEL == 0);
```

Comment:
ok - added comment:

    //  Serial access to globaldata->averageSpeed is essential to
    //  avoid data races.  Access to this variable is serialized
    //  in any case, so parallel computation wouldn't gain much,
    //  if anything.